shell/imports/server/installer.js

```diff
@@ -131,46 +131,42 @@ if (!Meteor.settings.replicaNumber) {
   });
 }
 
-doClientUpload = (stream) => {
-  return new Promise((resolve, reject) => {
-    const id = Random.id();
-
-    const backendStream = globalBackend.cap().installPackage().stream;
-    const hasher = Crypto.createHash("sha256");
+export function readPackageFromStream(stream, backend, progress = function() {}) {
+  // readPackageFromStream reads an spk pacakge from stream and uploads it into
+  // backend.
+  //
+  // The optional progress callback will be invoked each time a chunk is written,
+  // which is useful e.g. for updating progress bars. The length of the chunk is
+  // passed as an argument.
+  //
+  // Returns a promise which resolves to an object with fields:
+  //
+  // info: the return value of PackageUploadStream.saveAs().
+  // packageId: The package id of the uploaded package.
+  const backendStream = backend.cap().installPackage().stream;
+  const hasher = Crypto.createHash("sha256");
 
-    stream.on("data", (chunk) => {
-      try {
+  return new Promise((resolve, reject) => {
+    pipeline(stream, new Writable({
+      write(chunk, _encoding, callback) {
         hasher.update(chunk);
-        backendStream.write(chunk);
-      } catch (err) {
-        reject(err);
-      }
-    });
-
-    stream.on("end", () => {
-      try {
-        backendStream.done();
-        const packageId = hasher.digest("hex").slice(0, 32);
-        resolve(backendStream.saveAs(packageId).then(() => {
-          return packageId;
-        }));
-        backendStream.close();
-      } catch (err) {
-        reject(err);
-      }
-    });
-
-    stream.on("error", (err) => {
-      // TODO(soon):  This event does't seem to fire if the user leaves the page mid-upload.
-      try {
-        backendStream.close();
+        backendStream.write(chunk).then(() => {
+          callback();
+          progress(chunk.length);
+        });
+      },
+    }), (err, _val) => {
+      if(err) {
         reject(err);
-      } catch (err2) {
-        reject(err2);
+        return;
       }
+      backendStream.done();
+      const packageId = hasher.digest("hex").slice(0, 32);
+      backendStream.saveAs(packageId)
+        .then((info) => resolve({info, packageId}), reject)
```

Comment:
Callback-oriented code always makes me nervous that errors might not be handled. You could convert more of this into async/await like so:

```suggestion
      resolve((async () => {
        backendStream.done();
        const packageId = hasher.digest("hex").slice(0, 32);
        let info = await backendStream.saveAs(packageId);
        return {info, packageId};
      })());
```